src/client/Client.cc

```diff
@@ -1729,6 +1729,12 @@ void Client::connect_mds_targets(mds_rank_t mds)
 	mdsmap->is_clientreplay_or_active_or_stopping(rank)) {
       ldout(cct, 10) << "check_mds_sessions opening mds." << mds
 		     << " export target mds." << rank << dendl;
+
+      auto session = _get_or_open_mds_session(rank);
+      if (session->state == MetaSession::STATE_OPENING ||
+          session->state == MetaSession::STATE_OPEN)
+        continue;
```

Comment:
How is the issue [1] solved ? Also, with this code in-place, do we even end up in the scenario of handling session already opened in mds as done in commit fa18e28ea ?

[1] https://tracker.ceph.com/issues/53911

Reply:
The third commit is just a improvement, not fixing any of these two issues here.

For [1] issue, since the clients may send new open session requests after the session already opened, currently the MDS will just discard these request without telling the clients what has happened. So the clients will wait forever.

The second commit https://github.com/ceph/ceph/commit/fa18e28eaea47a1f9f8bc0e59ee0223c6757b80d  will try to tell clients what has happened for this session.

Reply:
> > The third commit is just an improvement, not fixing any of these two issues here.
> > For [1] issue, since the clients may send new open session requests after the session already opened, currently the MDS will just discard these requests without telling the clients what has happened. So the clients will wait forever.
> 
> With this fix, clients won't send new open session requests if the session is already opened right ? 

No, even without this fix the client won't do that too.

> You mean to say the older clients without this fix would do that ? I guess I am missing something.
> 

There has one case that for example:

if clientA try to open the session, but for some reason the connection was reset, then the socket connection will be broken, but leaving the session in open state in MDS, while the clientA will close the session. So next time when the clinetA want to send any new request, such as open file request it will try to open the session first and move the client file open request into wait list, then this time when opening the session, the MDS will just discard the session open request.

So the client doesn't know what has happened and keep waiting.....



> > The second commit [fa18e28](https://github.com/ceph/ceph/commit/fa18e28eaea47a1f9f8bc0e59ee0223c6757b80d) will try to tell clients what has happened for this session.